Expose OIDC tokens

velochy opened the pull request:
## Describe your changes

Store the  OIDC auth response in a cookie and expose them to users for API calls

Currently, from user perspective, `st.user` gets an extra field called "tokens" that is a dict of all the tokens that were made available - most often the triplet of acces_token, id_token, refresh_token

This is useful for multiple use cases (see the issues below).

To achieve (a) reliably, the tokens are stored in a separate cookie from user info. This is because cookies are limited to be <4096 bytes and I ran into that limit in my own testing already if I combined the two dictionaries. Additionally, both cookies can now be chunked so as a by-product, this now supports login systems with larger userdata.

I have tested it to work in my own use case and it seems to be performing as intended. But it is just one use case and I may well be missing some others. so any proposals to improve this are very welcome. 

Originally, also contained st.user_refresh command, but that got it's own PR: https://github.com/streamlit/streamlit/pull/12696

## GitHub Issue Link (if applicable)

Closes #10378 

Exposing access_token: https://github.com/streamlit/streamlit/issues/10378
Too large usedata for current login system: https://github.com/streamlit/streamlit/issues/12518

Need to provide id token hint at logout: https://github.com/streamlit/streamlit/issues/12144#issuecomment-3172828767 -> PR https://github.com/streamlit/streamlit/pull/12693
Need for user refresh: https://github.com/streamlit/streamlit/issues/12043 -> PR https://github.com/streamlit/streamlit/pull/12696

## Testing Plan

- Unit tests present
- Manually tested to work for my own use case

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

velochy committed:
Create a separate cookie for auth tokens and make them available in st.user

velochy commented:
@vdonato Thank you for getting the logout PR across the finish line yesterday :)
I'm hoping you'll help me with this one as well, as it is kind of a follow-up.

velochy commented on lib/streamlit/user_info.py:
```diff
@@ -353,6 +353,56 @@ def logout() -> None:
         context.enqueue(fwd_msg)
 
 
+AUTH_REFRESH_ENDPOINT: Final = "/auth/refresh"
```
Analogous to AUTH_LOGIN_ENDPOINT defined there and also not used anywhere else so keeping as-is for consistency

vdonato commented:
Hey @velochy, happy to assist with this PR as well!

Heads up that I do expect to be relatively busy for the next week or so, so it might take me a bit of time to get around to giving this a review. Feel free to ping me to remind me if you haven't heard back in a week's time.

velochy commented:
No prob @vdonato . I'll also be taking a trip and will be away from the computer for a few days. 
And while this would be a useful feature for my company, it's not a blocker for anything.

lukasmasuch commented:
I added the `status:needs-product-approval` since adding a new command - `st.refresh_user` - will need product approval. (cc @jrieke). Also, I think @jrieke already wrote a spec some time ago related to exposing the access token.

velochy commented:
@jrieke would love to see the spec and build to match it if you had a plan already.

Also @vdonato this is the gentle reminder nudge you asked me to provide :)

sfc-gh-jrieke commented:
Hey, I'm OOO for the next 1.5 weeks but will take a look afterwards!

velochy commented:
@sfc-gh-jrieke gentle reminder now that you are hopefully back.

While you were away, https://github.com/streamlit/streamlit/pull/12182 got reverted, as users were experiencing issues and for [at least one user](https://github.com/streamlit/streamlit/issues/12144#issuecomment-3172828767), it would seem the only way to fix it requires keeping track of id_token, which is provided by this PR. So it would be really nice to get this PR done soon-ish so I can fix the logout one and get that re-merged as well.

jrieke commented:
Hey @velochy, thanks for the wait! 

First of all, let me say these are things we really want (and we know many users ask for them, especially the access token). Unfortunately, the engineer who originally worked on authentication is no longer with Streamlit, so we currently have a pretty limited understanding of all the details in the team, which will make things slower. Plus, changes to this feature will need to go through an internal review with our security team, which again makes it a bit more complicated. So please bear with us! :D

I read up on your proposed changes and have a few questions for understanding – would be great if you could try to explain them in simple terms since I'm not an expert on OIDC:

1. Why do we need to expose the `id_token` in `st.user`? I understand there's this issue with having to provide the ID token at logout for specific providers (https://github.com/streamlit/streamlit/issues/12144#issuecomment-3172828767) but isn't that something we'd need to do behind the scenes in `st.logout` instead of the developer manually doing that? 

2. Exposing the `access_token` in `st.user` is something we probably want to do but I _think_ it might make sense to gate that behind a parameter (e.g. `st.login(..., expose_access_token: bool = False)`) to make it more secure (given that most users will probably not need the access token). Do you have any opinion on that? Also noting this will definitely need a review from our security team. 

3. Why do we need to expose the `refresh_token` in `st.user`? Isn't that something that would anyway be handled behind the scenes, even with your `st.refresh_user` function? As far as I understand it, exposing the refresh token poses a significant security risk since you can use it to get new access tokens, right? (Of course, if devs 100% understand all these tokens and handle them responsibly, there's no risk; but given that our users are mostly data-focused personas who don't have a great understanding of auth, I'd lean towards only exposing what's absolutely necessary.)

4. Is there a way we can refresh the user info/access token automatically behind the scenes instead of having to manually call `st.refresh_user`? I read your issue #12043 and see that you want to specifically update the user info manually but I think for most use cases, devs will probably just want to have a fresh access token when it expires, right? I think eventually I could imagine both an automatic refresh + a manual command to refresh whenever you need, just trying to understand what's possible and what the priorities are. Also, just want to note already that I'm not sure how well `st.refresh_user` fits into our API, so if we do want a manual command for this, we'd definitely need to think through naming and whether we make this a top-level command or can somehow integrate it more nicely with `st.user` or `st.login`.

velochy commented:
> Hey @velochy, thanks for the wait!

No worries. Summer is vacations time, and I was also largely away myself :)

> First of all, let me say these are things we really want (and we know many users ask for them, especially the access token). Unfortunately, the engineer who originally worked on authentication is no longer with Streamlit, so we currently have a pretty limited understanding of all the details in the team, which will make things slower. Plus, changes to this feature will need to go through an internal review with our security team, which again makes it a bit more complicated. So please bear with us! :D

I'll do my best. I currently have a very hacky workaround for this with two otherwise equivalent oauth confs, one with prompt="login" and one without, that in a roundabout way almost achieves the desired behavior. But I would really like to clean up my code and the only way to do that is... well to get this and the login PR merged to Streamlit. So I'm moderately incentivized, but it's not urgent. 

> I read up on your proposed changes and have a few questions for understanding – would be great if you could try to explain them in simple terms since I'm not an expert on OIDC:
 
> 1. Why do we need to expose the `id_token` in `st.user`? I understand there's this issue with having to provide the ID token at logout for specific providers ([st.logout() doesn't redirect to correct Cognito ESE #12144 (comment)](https://github.com/streamlit/streamlit/issues/12144#issuecomment-3172828767)) but isn't that something we'd need to do behind the scenes in `st.logout` instead of the developer manually doing that?

Technically we can leave id_token behind the scenes, but as APIs are different, I can easily see a scenario of some users needing that instead/alongside their access_token. So at least to me  it makes sense to expose both in the same place. Id token is relatively safe to expose, as it is basically the user object (current user object is basically extracted from it), just signed by the oauth provider - so it adds less risk than exposing the access token as most of the info contained in it is already in the object.

> 2. Exposing the `access_token` in `st.user` is something we probably want to do but I _think_ it might make sense to gate that behind a parameter (e.g. `st.login(..., expose_access_token: bool = False)`) to make it more secure (given that most users will probably not need the access token). Do you have any opinion on that? Also noting this will definitely need a review from our security team.

I will defer to your security team on this as I'm not a cybersecurity expert (fun fact: I do have a PhD in cryptography, but a) it was 15 years ago and b) I was heavily on the theory side, so my practical security knowledge was always flaky at best). Basic googling told me that storing the token in an encrypted and only server-accessible cookie is pretty standard practice though, so as best I can tell, it should be ok to do for all users. 

> 3. Why do we need to expose the `refresh_token` in `st.user`? Isn't that something that would anyway be handled behind the scenes, even with your `st.refresh_user` function? As far as I understand it, exposing the refresh token poses a significant security risk since you can use it to get new access tokens, right? (Of course, if devs 100% understand all these tokens and handle them responsibly, there's no risk; but given that our users are mostly data-focused personas who don't have a great understanding of auth, I'd lean towards only exposing what's absolutely necessary.)

The way I see it: we still need to store it, even if we don't expose it. And as long as we store it, anyone writing a streamlit app can still access the cookie that stores it. So the only thing exposing it changes is making it slightly more likely to leak by accident (i.e. someone posting the user object somewhere without filtering for fields). 

Which, now that I think about it, is not an insignificant risk. So yes. You are right in that refresh_token should probably not be exposed in the user object. Again - if someone really wants or needs it, they can read the cookie and access it, but it would require very deliberate effort an control of the streamlit application code. 

> 4. Is there a way we can refresh the user info/access token automatically behind the scenes instead of having to manually call `st.refresh_user`? I read your issue [Option to refresh user info for user authenticated via st.login #12043](https://github.com/streamlit/streamlit/issues/12043) and see that you want to specifically update the user info manually but I think for most use cases, devs will probably just want to have a fresh access token when it expires, right? I think eventually I could imagine both an automatic refresh + a manual command to refresh whenever you need, just trying to understand what's possible and what the priorities are. Also, just want to note already that I'm not sure how well `st.refresh_user` fits into our API, so if we do want a manual command for this, we'd definitely need to think through naming and whether we make this a top-level command or can somehow integrate it more nicely with `st.user` or `st.login`.

As you correctly noted, for my use case, i.e. refreshing the user object after the user has changed their settings like language (which I don't think is an uncommon use case), a manual refresh is definitely needed. But it can easily be moved out of sight, for instance to  st.user.refresh() instead of the current st.refresh_user().

As for automatic refreshing of access tokens, we could start every session with a refresh if we have a cookie present (which is basically what I plan to do in my app), but there are legitimate cases where people might not want that (i.e. to keep login periods short and force a relogin every day or so for security reasons), so we would then need to add a configuration variable to control this behavior. And I think we would both prefer to not create another rabbithole on an already complex PR ;)

velochy commented:
> (e.g. st.login(..., expose_access_token: bool = False))

On this proposal, it is worth noting that reacting to st.login parameters is very cumbersome, as they would need to be passed through the OIDC authentication process via ?state parameter which is currently only used to avoid replay attacks. 
So while doable, it would add A LOT of additional complexity, and considering you told me the person in charge of this part of code is no longer with you, this is something you are likely looking to avoid. 

A far easier solution (at least on the technical side) would be to have a secrets.toml flag controlling this behavior, as that object is easily accessible everywhere and usually already accessed in the relevant code parts. So having something like
`expose = "acess,id"`
that controls which tokens are exposed with a default value of an empty string might be the most reasonable solution

pallagj commented:
To add some concrete context, especially around the need for the `access_token`, here’s my primary use case. This should help clarify why exposing it is not just a convenience but a **critical feature** for building more integrated and secure applications with Streamlit.  

In the application I'm developing, we use **Keycloak** as our identity provider (IdP). A key feature we want to offer is the ability for certain privileged users (e.g., team managers) to create and manage other users directly from within the Streamlit application's UI.  

---

### Without API

There’s also a big **administrative overhead**. For the app to function, a user record must exist both:  

1. In **Keycloak** (for authentication).  
2. In the **Streamlit app’s database** (for app-specific data and settings).  

Currently, this means:  
- Create the user in Keycloak.  
- Copy the unique Keycloak user ID.  
- Manually create a corresponding user in the Streamlit app’s DB and paste in the Keycloak ID.  

This **dual-entry process** is tedious, error-prone, and does not scale.  

---

### The Ideal (and Secure) Workflow This PR Enables  

1. A manager logs into the Streamlit app via OIDC with their Keycloak account.  
2. The Streamlit backend receives and securely stores the user’s `access_token` (as proposed in this PR). This token contains the permissions and roles assigned to that specific manager in Keycloak.  
3. Inside the app, the manager navigates to a user management page and fills out a form to create a new user.  
4. On submission, the Streamlit backend takes the manager’s own `access_token` and uses it to make a secure API call to the **Keycloak Admin API** to create the new user.  

**Why this is secure:**  
- Operates under the **Principle of Least Privilege**.  
- Success depends entirely on the permissions granted to the logged-in manager.  
- If they have the right role, Keycloak allows the action; if not, it’s denied.  
- No additional or long-lived credentials are required.  

---

### The Current (Insecure and Cumbersome) Workaround  

Without access to the logged-in user’s token, the process looks like this:  

- I must create a powerful **“service account”** or admin user in Keycloak with broad permissions.  
- Store the credentials (e.g., client ID and secret, or username/password) for this admin user in `secrets.toml`.  
- When a manager tries to create a user, the backend authenticates with Keycloak in the background using these stored admin credentials.  
- It then uses this **all-powerful admin token** to call the Keycloak Admin API.  

**Problems with this approach:**  
- Requires storing long-lived, high-privilege credentials in app configuration.  
- Bypasses the granular, role-based permissions of the actual user.  
- Increases attack surface and creates unnecessary risk.  

---

### Why This PR Matters  

By exposing the `access_token`, this PR allows developers to:  

- **Eliminate** the insecure service-account workaround.  
- Build integrations authenticated by the actual end-user’s identity and permissions.  
- **Automate user provisioning**, avoiding manual syncing between systems.  
- Dramatically improve the **security posture** and reduce admin overhead.

pallagj commented:
Building on the valid security concerns raised, perhaps there's a middle ground that provides both developer flexibility and enhanced, user-centric security.

What if accessing the token required explicit user consent via a pop-up?

Instead of the token being directly available in the `st.user` object, a developer would have to call a function to retrieve it. For example:

```python
# The developer requests the token
access_token = st.user.get_access_token()
```

When this function is called for the first time in a session, Streamlit could present a modal dialog to the end-user:

> "This application is requesting permission to use your authentication token to perform actions on your behalf. Do you approve?"
>
> [Allow] [Deny]

The `get_access_token()` function would only return the token if the user clicks "Allow." This function could also handle the automatic token refreshing in the background.

**Advantages of this approach:**

- **Puts the User in Control:** It makes token access an explicit, user-approved action. This prevents an application from silently using the token in the background and ensures the user is aware of what permissions they are granting.

- **Maintains Flexibility:** Unlike a fully abstracted `st.authenticated_request()` function, this approach still gives the developer the actual token. This is crucial for using complex third-party SDKs or libraries that require the token to be passed directly.

- **Balances Security and Functionality:** It directly addresses the risk of accidental token exposure by making the access intentional and auditable from the user's perspective, while still enabling the powerful integration use cases that motivated this pull request.

velochy commented:
@jrieke have you had a chance to think on it, and does anything else need clarifying?

Based on your (very good) questions and the technical limitations, my proposal would be to not expose any tokens by default, but to add an "expose" flag to secrets.toml that allows user to specify what they need (access, id) and then those get added to the st.user.

In terms of refresh, moving it from st.refresh_user() to st.user.refresh() seems pretty sensible and straightforward, as UserInfoProxy is already a class due to needing to override default setters on dict. 

If this makes sense to you as well, I'd be happy to make the changes.

velochy commented:
@jrieke @vdonato there has been no reply here over a week, so just a gentle reminder - I'm happy to make any changes as long as I know which ones :)

jrieke commented:
Sorry, didn't get back to this yet since we have annual planning going on. Will have a look in the next few days!

velochy commented:
@jrieke another gentle reminder :)

jrieke commented:
Ok finally getting back to this, was out sick :/ 

> Based on your (very good) questions and the technical limitations, my proposal would be to not expose any tokens by default, but to add an "expose" flag to secrets.toml that allows user to specify what they need (access, id) and then those get added to the st.user.

Agree on this point. I'd say we call this parameter `expose_tokens` for now and make it accept both a string or a list, so you can do any of the following:

```toml
expose_tokens = "id"
expose_tokens = ["id"]
expose_tokens = "access"
expose_tokens = ["access"]
expose_tokens = ["id", "access"]
```

As discussed above, I'd leave out the refresh token for now due to security. Btw I do also want to allow passing any parameters from the `[auth]` section in `secrets.toml` to `st.login` at some point (to solve #10543) but I think for now it's fine if we keep this in the toml file. 

The exposed tokens should then go into `st.user.tokens`, which should (just like `st.user` itself) be a dict-like object that can be accessed both via dot and key notation, so you can e.g. do:

```python
st.user.tokens.id
st.user.tokens["access"]
```

If no token is exposed, `st.user.tokens` should be an empty dict I think. 

Let me know if that sounds good! I'll also talk with our security team to get a first assessment. 


> In terms of refresh, moving it from st.refresh_user() to st.user.refresh() seems pretty sensible and straightforward, as UserInfoProxy is already a class due to needing to override default setters on dict.

Hmm I'm not sure about this one. The annoying part is that we decided to put `st.login` and `st.logout` at the top level instead of under `st.user.login` and `st.user.logout` because we didn't really consider that there could be additional functions we'd need. Now it seems pretty annoying to have the original functions at the top level, but have `.refresh` under `st.user`. At the same time, it's also annoying to put the refresh function (which will be pretty niche I think) at the top level. Since I think this is relatively independent from the access token PR, could we maybe split that out into a separate PR, so we can get the access token work merged and then think about the refresh function?

velochy commented:
@jrieke I believe your response provides a pretty good spec to work towards and seems both sensible and relatively straightforward to implement. 

I have a work deadline end of next week that will likely require most of my coding focus, but I should be able to get it done the week after that if all goes well.

Regarding the refresh: you are right it is enough of a separate concern to make sense moving to a new PR (and I will), but as it is still a feature I very much need, I would still ask you think on how you want it solved. I understand your dilemma, but at some point a call just needs to be made one way or another, and I'd very much prefer if you could do it sooner rather than later ;) As for it being very niche - it is definitely less needed than st.login or st.logout, but I don't believe my use case of changing user info and then wanting it to be updated in st.user is very far out there. And for my two cents, I don't think st.user.refresh() would be that big of a pattern break, but you could potentially ameliorate it by mirroring st.login and st.logout on st.user as well (so they are available both top level and on user) and change documentation to talk about them in the context of st.user but just note they are mirrored on top level for convenience

velochy committed:
Changes based on https://github.com/streamlit/streamlit/pull/12044#issuecomment-3310054404

velochy commented:
Made the changes as agreed:

* Removed everything regarding refresh tokens for a separate PR
* Added a filter to only expose user specified tokens
* Made st.user.tokens dict-like (this turned out to be somewhat verbose, unfortunately)

One important note worth pointing out is that while the code currently exposes only the tokens the user specifies, it still stores -all- of them in a cookie. This is done so we could still internally use them as needed. Most notably, we will need the id cookie to do proper logout (https://github.com/streamlit/streamlit/issues/12144#issuecomment-3172828767)  and will need the refresh token for the refresh functionality down the line. As the cookie is properly secured (server-side and domain-restricted), this should not cause any security issues. 

I will create the new PR-s for logout and refresh on top of this branch and rebase them as needed so they would be ready once this gets merged.  I will hopefully get around to creating both PR-s tomorrow

velochy commented:
Both PR-s now created: Logout at https://github.com/streamlit/streamlit/pull/12693 and Refresh at https://github.com/streamlit/streamlit/pull/12696

AlvarezEtc commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -72,33 +72,40 @@ def initialize(self, base_url: str) -> None:
         if len(serialized_cookie_value.encode()) > 4096:
             _LOGGER.error(
-                "Authentication cookie size exceeds maximum browser limit of 4096 bytes. Authentication may fail."
+                "Cookie size exceeds maximum browser limit of 4096 bytes. Authentication may fail."
```
When storing cookie, sometimes they don't fit and it fails silently. It measures the cookie size before base64 encoding and before adding other info (sizes, checksum, timestamp). For example: 
`{"access_token": "1234", "id_token": "5678"} ` is 45 characters. 
But the actual cookie would be something like `2|1:0|10:1759839146|22:_streamlit_user_tokens|60:eyJhY2Nlc3NfdG9rZW4iOiAiMTIzNCIsICJpZF90b2tlbiI6ICI1Njc4In0=` which is 109 characters, and the actual encoded data is 60 characters instead of 45.

AlvarezEtc commented:
I have been testing this PR as I am interested in the feature that it will provide. While testing I found an issue with calculation of the cookie size limit. This affects the handling of any authentication cookie, but while storing tokens it is quite issue to reach the limit. Apart from fixing the calculation I would consider an alternative, such as splitting cookies. Please see the comment in code.

velochy commented:
@AlvarezEtc thank you for taking the time to test it and finding this issue.

To be honest, I was worried this might happen under some circumstances, but I was reluctant to add the extra complexity of cookie splitting "just in case". But you are right it makes sense to be rather safe than sorry.

I see the AI also has some fairly good suggestions. Also, seems tests fail for some versions due to typing issues. I'll try to take some time in the weekend to incorporate all the feedback at once.

velochy committed:
Added cookie chunking and other minor fixes based on AI comments

velochy commented:
@AlvarezEtc pushed a change that adds chunking. Because of needing to be backwards-compatible with old tornado, the code is rather awkward in places, but I did my best to keep it clean and readable.

The chunking has the added benefit of handling larger user data. Just to test, I copied 10k characters of Lorem Ipsum into a key in our OIDC provider for a test user, and sure enough, this makes the current streamlit login fail, whereas the updated version works as it should.

AlvarezEtc reviewed (approved):
LGTM

velochy commented:
Had to rebase again as there was a tiny trivial merge conflict.  Would be good if anyone could re-approve the test workflows to see if anything else needs fixing or cleaning @AlvarezEtc @jrieke

AlvarezEtc commented:
Hi, I have tested the flows to save and then to retrieve the tokens. It works as expected.

jrieke commented:
Fantastic, thanks so much @velochy! I approved the workflows. We're in quarterly planning right now and I made it a P0 priority that we assign an engineer to this PR to help you get it across the finish line and handle all the security stuff with our security team. We're currently figuring out who this will be but I hope by the end of this week we're a bit more organized. I hope then this will move forward a bit faster than when I have a look in my spare time 😅 

Will start looking at your other PRs after I'm back from vacation next week!

velochy commented:
Thank you @jrieke . Help would very much be appreciated. Tests fail again because of some `ty check` issue that I fail to reproduce on my computer, but which should be very easy to figure out for anyone who has the exact required dev env properly set up. 

And yes, a security review is definitely warranted for a feature like this. 

It should be pretty calm couple of weeks for me, so I can handle any ongoing work with a few days delay at most in case anything else is needed from my end :)

velochy commented:
@jrieke two weeks have again passed with no movement on this, with 3 months since it was first opened.
Is this normal? My hope was things move on the scale of weeks, not months.

I do have 2 follow-up PR-s to this, and I was hoping to get all of them done within this calendar year. Is that still realistic?

kmcgrady commented:
Hey @velochy! Just wanted to let you know we haven’t forgotten about this! I’m planning to review the PR and start the necessary conversations with our security team this week. I can’t commit to a timeline yet, but once I’ve had a chance to dig in, I’ll share a clearer sense of next steps.

> Is this normal? My hope was things move on the scale of weeks, not months.

That’s a totally fair question. In open source, especially when it’s tied to a larger company like Snowflake, things can sometimes move more slowly than we’d like. Our small team balances community contributions alongside internal goals, reviews, and planning cycles, which can make timing a bit unpredictable.

That said, your contribution is genuinely appreciated and valuable. These kinds of PRs push Streamlit forward in important ways, and we really don’t want you to feel discouraged. Thank you for your patience and for taking the time to invest in improving the project. It really does make a difference! 😊

velochy commented:
@kmcgrady thank you for the response. As I've mostly worked in startups myself, this feels off but I do understand things slow down as companies grow. 

Let me know if you need any help with anything.

kmcgrady commented on lib/streamlit/web/server/browser_websocket_handler.py:
The type error is due to the `user_info` being the wrong type.

This could be easily changed to `user_info: dict[str, str | bool | None | dict[str, str]] = {}`

That being said. I am wondering whether we should be considering a TypedDict and see if we can get a little more specific.

velochy replied:
Currently went with your proposed fix, and clarified filtered_tokens: dict[str,str] below just in case as well.
Not sure what you mean by TypedDict approach, as I have not used python typing too much in my own projects.

kmcgrady commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +173,177 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+    if main_chunk is None:
+        _LOGGER.exception("Missing main cookie (chunk 0) for '%s'", cookie_name)
+        return None
+    chunks.append(main_chunk.decode("utf-8"))
```
We shouldn't need to decode here, right?

kmcgrady replied:
Ah, see my below comment. It seems like we can simplify here?

velochy replied:
indeed. Good catch. Not sure what i was thinking. Might have not figured out the b"" and worked around it...
Anyway, your proposal is clearly superior.

kmcgrady commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +173,177 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+            return None
+        chunks.append(chunk_value.decode("utf-8"))
+
+    reconstructed_value = "".join(chunks)
```
Can't this just be `reconstructed_value = b"".join(chunks)` which will join the bytes rather than decoding/encodin?

zahhar commented:
> That said, your contribution is genuinely appreciated and valuable. These kinds of PRs push Streamlit forward in important ways, and we really don’t want you to feel discouraged. Thank you for your patience and for taking the time to invest in improving the project. It really does make a difference! 😊

Dear @kmcgrady - apologies for stepping into discussion without writing any single line on code here. Let me be user's voice in this thread ;-) - I am monitoring this PR for the last 2 month, because my project literally depends on it. What @velochy accomplished alone – is a huge piece of work (Aitäh, Margus!) that is long waited by hundreds of Streamlit builders. 

I am observing this discussion from September, and had hope this PR (alongside two others that address logout) will be part of v1.50, then – 1.51, and how I still hope it to be part of 1.52. It would be very pity if this amazing feature will be stalled because of merge conflicts or other issues preventing it from being released.

Thank you & your colleagues for your time and energy, please-please-please, support this feature to become part of Streamlit!

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -378,6 +378,124 @@ def _get_user_info() -> UserInfo:
 
+class TokensProxy(Mapping[str, str]):
+    """
+    A read-only, dict-like object for accessing exposed tokens from the\
```
nit: remove trailing slash

kmcgrady commented on lib/streamlit/user_info.py:
```diff
@@ -378,6 +378,124 @@ def _get_user_info() -> UserInfo:
+
+    def __len__(self) -> int:
+        return len(self._tokens)
+
```
This is fine, but wanted to note we do a bunch of similar things here, and a part of me thinks we should consider consolidating at some point.

1. https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/util.py#L82
2. https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/secrets.py#L146
3. https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/state/query_params_proxy.py#L27
4. https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/state/session_state_proxy.py#L69
5. The UserInfoProxy below.

velochy replied:
I looked at the UserInfoProxy and was a bit surprised you were not inheriting from a read-only-dict class that handles the boilerplate. But as an external contributor, I assumed you had a reason.

I agree it's a code smell, but this kind of refactoring is generally best left for internal teams for a reason. At minimum, it would/should be a separate PR. I'd rather just keep things as they are for now if that's ok?

kmcgrady replied:
That's okay. Likely we need some generic proxy that's outside of scope, but wanted to highlight this for future readers.

kmcgrady commented on lib/streamlit/web/server/browser_websocket_handler.py:
```diff
@@ -171,6 +186,20 @@ def open(self, *args: Any, **kwargs: Any) -> Awaitable[None] | None:
+                            if token_key in all_tokens:
+                                filtered_tokens[token_type] = all_tokens[token_key]
+
+                        user_info["tokens"] = filtered_tokens
```
Would this be a good option to put `TokensProxy(filtered_tokens)`? so that you don't need to update the TokensProxy below?

velochy replied:
Possibly? But by the same logic, would it not make sense to create the UserInfo object right then and there?

I'm operating by analogy (as you do in an unfamiliar codebase) and I was assuming there was a reason. For instance, the user_info object might need robust serialization which standard dict is better suited for than classes etc. 

It feels safer to leave it as it is, but I can change it if you want me to.

kmcgrady replied:
Ah that makes sense. I think we can separate this as we talk about uniting the proxy abstraction

kmcgrady commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +173,177 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+    actual_cookie_size = len(cookie_name) + 1 + len(signed_value) + 30
+
+    # Check if cookie needs to be split
+    if actual_cookie_size > 4096:
```
Could we have this as a constant here `MAX_COOKIE_BYTES`?

velochy replied:
Of course. Fixed it

kmcgrady commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +173,177 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+
+    # Cookie format: "name=value; HttpOnly; Path=/"
+    # Approximate additional bytes for cookie attributes (~30 bytes)
+    actual_cookie_size = len(cookie_name) + 1 + len(signed_value) + 30
```
What is the +1 here?

I guess constants would be nice here instead of the comment
`len(cookie_name) + SOME_CONSTANT + len(signed_value) + APPROX_COOKIE_ATTR_SIZE`

velochy replied:
Don't remember to be honest. Probably a separator or end marker of something.
Currently just added it to the 30 in the end and set APPROX_COOKIE_ATTR_SIZE = 31

kmcgrady reviewed (commented):
Hey @velochy I did my initial review and things look overall good. Most of the feedback are asking questions and smaller nits.

I spoke with security, and they are aware they need to review it. I will follow up with them next week, if I do not hear back.

velochy commented:
@kmcgrady fixed all of the small things except https://github.com/streamlit/streamlit/pull/12044#discussion_r2505849884 which feels like it warrants a separate PR

kmcgrady commented:
Awesome. Taking another look, and it looks very reasonable. Looks like some tests/more linting failed. I might push up a change that fixes that, but that seems relatively minor.

I meet with Security tomorrow to start the conversation. I'll report back the results. 

For other readers following this thread, this PR does generate a [wheel file](https://github.com/streamlit/streamlit/actions/runs/19191882868?pr=12044) you can use (see the artifacts) as a way to try out and test it yourself. More than welcome to use it in your work (at your own risk of course).

kmcgrady commented:
Hey @velochy Security got back and overall this is good. They came back to me giving the overall go-ahead. They have a couple comments that I can add to this PR directly to get this merged. Let's hope by end of this week! 😄

kmcgrady commented:
One last thing @velochy , usually all the feature work we do is written with a spec in mind. We have started hosting public specs, so that we can have more discussions in the open. I've been trying to decipher the "spec" inside all the discussions in this PR, but it was a little hard to follow. So I used AI to synthesize it and create a spec. You can [see it in this PR](https://github.com/streamlit/streamlit/pull/13044). Would you mind just taking a look and see that it aligns with what you implemented/believed? @jrieke is also going to post any comments as well.

kmcgrady commented:
Hi @velochy Thanks for the patience here. I have to be the bearer of some bad news: I won't be able to get this merged in time for release 1.52 (slated for next week).

I started addressing the type issues and unfortunately fell down a rabbit hole of complexity there. I also had to prioritize a requested fix from Security regarding cookie chunking behavior (removing the count cookie to align with other frameworks).

It is a holiday week here in the US, so I will be taking a break and will resume this when I am back next week. I know this delay is frustrating, but I want to reiterate the value of this PR. I am doing my best to stay on top of it to ensure it gets merged while balancing my other priorities.

kmcgrady committed:
Address feedback from security

lukasmasuch commented:
@cursor review

kmcgrady committed:
Fix feedback and redact tokens on st.json

kmcgrady commented:
> However, the incorrect use of _LOGGER.exception() should be fixed before merging as it will produce confusing log output without actual exception tracebacks.

This violates the ruff recommendation, so I think we should choose one, but will keep exception for now.

jrieke commented:
Finished and merged the spec now! @velochy as you pointed out, this should already conform with your implementation, but just for safety would be good if you could give this a final look and double check :) 

Two more points that came to my mind while finishing the spec:
- Did you test that this works on Community Cloud? I guess it should given that normal auth works there as well, but would be good to do a quick smoke test to see that nothing changed. 
- Are we collecting telemetry metrics for this in some way? Would be good to either track how often the `expose_tokens` parameter is used (even though not sure whether we want to track things from `secrets.toml`) or how often `st.user.tokens` is accessed. This helps us figure out how often this feature is used in practice and how we should prioritize furutre improvements. @kmcgrady and @lukasmasuch can you help with this point?

lukasmasuch commented:
I believe we can just annotate the `token` property function on `st.user` with the metrics decorator

kmcgrady committed:
Add metrics

kmcgrady commented:
Confirmed that the changes conform to the spec. **This currently will not work on community cloud**. This is due to security and filtering out unexpected cookies. I have a change under review, and we can deploy the latest changes to work in production.

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -79,11 +88,36 @@ def get_secrets_auth_section() -> AttrDict:
+    - expose_tokens = ["id", "access"] -> ["id", "access"]
+    """
+    auth_section = get_secrets_auth_section()
+    expose_tokens = auth_section.get("expose_tokens")
```
**question:** If I'm reading this correctly, this only reads from the top-level `[auth]` config section. However, the docs in `TokensProxy` mentions that it can be set from `[auth]` and `[auth.<provider>]` (eg: `[auth.google]`) config blocks. Which one should be correct?

kmcgrady replied:
I believe having it for the provider does not make as much sense here (at least for now), but I will confirm with @jrieke We load the user info and token info on browser connection, and we don't store any context on the provider, so I expect that the `expose_tokens` to be a global. I think we can consider storing the provider in the cookie as well in order to determine which provider section to look into.

velochy replied:
Provider is indeed useful to store (logout needs it, for instance) but it makes sense to put that into userinfo rather than under tokens

jrieke replied:
Both fine to me! Might be a tiny bit nicer to be able to set it per provider, but if that's hard, totally fine to just keep it as a global setting and wait until someone creates an issue. I don't think that's a super common use case.

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +178,173 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+    """Split a large cookie value into multiple smaller cookies.
+
+    The main cookie always exists and contains the first chunk.
+    Additional chunks are stored as cookie_name_1, cookie_name_2, etc.
```
**issue:** This docstring seems to be out of line with what is actually implemented. It looks like if there are multiple cookies then the first cookie is just the count.

kmcgrady replied:
Will correct.

sfc-gh-bnisco commented on lib/streamlit/web/server/oauth_authlib_routes.py:
```diff
@@ -72,33 +73,71 @@ def initialize(self, base_url: str) -> None:
+            AUTH_COOKIE_NAME,
+            user_info,
+        )
+        set_cookie_with_chunks(
+            self._set_single_cookie,
+            self._create_signed_value,
+            TOKENS_COOKIE_NAME,
+            tokens,
+        )
```
**question:** It seems like we're always setting this now. Shouldn't we only set this if `expose_tokens` is configured?

velochy replied:
No. We will need the id token and refresh token internally for proper logout/refresh so the cookie should be set irrespective of expose_tokens. 

See the two other PRs linked above that are waiting for this to get merged.

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -144,6 +178,173 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+
+    # Maximum size for a single cookie chunk (conservatively set to allow for overhead)
+    # Overhead seems to be around 110 bytes, but depends on Tornado internals, so having a safe 3x margin
+    cookie_space = 3800 - len(
```
**question:** How sure are we about this `3800` magic number? Would it make more sense to actually compute this instead so that we are both more efficient and never going over the max size?

kmcgrady replied:
It's a 3X margin, but I think I have an idea to make this cleaner.

kmcgrady committed:
Address feedback

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -141,9 +180,216 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+    signed_value = create_signed_value_fn(cookie_name, serialized_cookie_value)
+
+    # Cookie format: "name=value" + COOKIE_ATTRIBUTES
+    actual_cookie_size = len(cookie_name) + len(signed_value) + COOKIE_ATTR_SIZE
```
**suggestion:** This should include `=` in the size calculation.

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -141,9 +180,216 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+    available_for_base64_value = available_for_signed_value - signing_overhead
+
+    # Convert from base64 space to raw value space (base64 has 4/3 expansion ratio)
+    chunk_size = (available_for_base64_value * 3) // 4
```
**question:** Do we need guard this potentially being a negative number?

kmcgrady replied:
You thinking just raising an error here? I figure if it's negative that means there's an error in the server signing (or our logic of sending the right function). Happy to add an error, but it might be a little overkill

sfc-gh-bnisco replied:
Yeah, I suppose raising an error makes most sense. I believe that it would output malformed tokens if we didn't raise here, which seems like a worse failure state.

sfc-gh-bnisco commented on lib/streamlit/user_info.py:
**suggestion:** These docstrings need to be updated because tokens are no longer ignored with this feature.

sfc-gh-bnisco commented on lib/streamlit/auth_util.py:
```diff
@@ -141,9 +180,216 @@ def generate_default_provider_section(auth_section: AttrDict) -> dict[str, Any]:
+        chunk_name = f"{cookie_name}_{i + 1}"
+        chunk_value = get_single_cookie_fn(chunk_name)
+        if chunk_value is None:
+            _LOGGER.exception("Missing chunk %d for cookie '%s'", i, cookie_name)
```
**suggestion:** I think this needs to use `i + 1`.

kmcgrady committed:
Address feedback

kmcgrady committed:
Fix minimum base64 space check to require at least 4 bytes

sfc-gh-kmcgrady commented:
@velochy I know it's been a while! But everything got approved and I successfully merged it in! Thank you again for your patience here!

jrieke commented:
Amazing, thank you so much everyone and especially @velochy! This is (I think) the most complex external contribution we ever got, so huge kudos to you. We'd love to send you some swag as a thank you, feel free to fill out the form here, and we'll get something nice on the way: https://forms.gle/RZCUTKCrg7aHyqH78

velochy commented:
> Amazing, thank you so much everyone and especially @velochy! This is (I think) the most complex external contribution we ever got, so huge kudos to you. We'd love to send you some swag as a thank you, feel free to fill out the form here, and we'll get something nice on the way: https://forms.gle/RZCUTKCrg7aHyqH78

Wow. Would not have thought this would be the case as I don't feel the contribution was particularly big.
Thank you to @kmcgrady and @jrieke  for getting it over the finish line. I see a fair bit of work still went into it after I thought I was done :)

I'll try to find some time over the holidays to rebase the other two PR-s and review them in light of the comments here. I really do feel both are important for the OAuth functionality to be fully usable. But yes, right now, the ball is back in my court :)

Meanwhile, happy holidays for everyone involved, and thank you all - @jrieke @kmcgrady @sfc-gh-bnisco @lukasmasuch @pallagj
